### ISSUE TITLE
Add support for HTTP/2 ‘push’

### ISSUE THREAD

**Issue by evyncke** [YangCatalog/deployment#114]
Please note that I am unsure to which this issue should be attached... Possible search and web_ui ?

As YC.O supports HTTP/2, let’s configure the server (or the front-end) to support PUSH requests to accelerate the site load (notably the huge script/style files).

HTTP/2 'push' allows a browser to start downloading important images, CSS, scripts, ... while still downloading the main page. In some cases, it allows for faster page display. It requires sending a "Link" HTTP header before sending the HTTP body.

See also: https://blog.cloudflare.com/http-2-server-push-with-multiple-assets-per-link-header/ for the syntax of the "push" part. Chrome developper console in the networking view can add a column for protocol (should be H2) and initiation (should be push or other).

**Comment by SlavomirMazurPantheon** [comment]
Hello @evyncke,
I recently made some changes in NGINX config files (see: https://github.com/YangCatalog/deployment/pull/115).
In addition, I did some research on the possibility of using HTTP/2 PUSH functionality.

Load time of yangcatalog homepage is around 3 - 5 seconds without caching. However, browsers already has a cache so with caching it is approximately 0.8s - 1.5s. It means that if you are accessing yangcatalog for the first time it takes quite a bit of time, but it is significantly less with next page visits.

Basically, I think HTTP/2 PUSH only makes sense for YANG Catalog if the user visits the site for the first time.
So I implemented it in this way for now. However, HTTP/2 PUSH is now working for YANG Catalog (I made initial NGINX pre-setup) so we are able to use and modify it in the future.

Regards,
Slavomir Mazur

**Comment by evyncke** [comment]
Currently, I have only IPv4 access to YC.O and the nginx configuration appears to enable HTTP2 only on IPv6.
Can you investigate to have IPv6 access again to YC.O ?

**Comment by SlavomirMazurPantheon** [comment]
HTTP2 is also enabled for IPv4.
IPv6 is now working properly.

**Comment by evyncke** [comment]
Thank you for re-activating IPv6.

HTTP/2 is indeed active now but not yet the pre-loading, i.e., using the "Link:" header (see blogpost above). When I see the large CSS/JS files (notably main*.js and style*.css) and the yang-catalog-header.png, the "Link:" would (slighly) increase the load speed even on the first access. (for the home page at least)

**Comment by SlavomirMazurPantheon** [comment]
I enabled HTTP/2 PUSH for two resources:
1. yang-catalog-header.png image -> as this is the largest resource loaded from backend
2. styles.css file

If you want to test it, first open [yangcatalog](https://yangcatalog.org/) in Incognito/Private window, then open developer tools (Ctrl + Shift + I) and clear cookies for yangcatalog.org in Application tab. Then go to the Network tab and hit Ctrl + Shift + R.

As I mentioned HTTP/2 PUSH is used only when first time visiting page (= we are testing if _session_ cookie is set).
However, as you mentioned I think that load speed has increased slightly - mostly due to pushing yang-catalog-header.png file.

**Comment by evyncke** [comment]
I confirm indeed the described behaviour. But, honestly, I do not understand why you are pushing it only on the first visit. If the content is already cached, as far as I know, the PUSH will not be accepted by the browser, so, nothing will be downloaded. So, it should not be harmful and would help for the case when the cache has expired and the cookie is still there.

Would it be possible to add the main*.js as well ? I appreciate that the 'random' part in the filename makes things more complex.

Anyway, already a good progress ;-) Thank you.

**Comment by SlavomirMazurPantheon** [comment]
Yes that 'random' part might be a problem, but I can try to find out if it is possible for such files as well (maybe some regex? ).
Also, I will once again check that part with pushing only for first visit.

I'll let you know when I find out something new.

**Comment by evyncke** [comment]
Just checked for the push this morning as the site is really slow... on the 'home' page, the 'link' header has disappeared hence no push at all.

On other pages, only the small style.css is pushed (no more the large PNG or even the huge 3 MB .js)

**Comment by SlavomirMazurPantheon** [comment]
Hello evyncke,

among other things, in recent weeks I have also looked at the possibility of using HTTP/2 PUSH in the YANG Catalog.
As I mentioned here: https://github.com/YangCatalog/yangcatalog-ui/issues/50

I changed functionality in the way that in NGINX config we only enable `http2_push_preload`.
This enables automatic conversion of preload links specified in the "Link" response header fields into push requests.

In this PR: https://github.com/YangCatalog/yangcatalog-ui/pull/54
I added creation of Link headers directly into index.html (so it was moved from NGINX into index.html file) so preload is initialized right after loading index.html -> which is always the first file loaded.
Preloading should start before main.*.js file is parsed.

Size of the yang-catalog-header.png was decreased by 53% in this commit: https://github.com/YangCatalog/yangcatalog-ui/commit/3ad8e061667ecb4c28c126f238907e4b663da698
This also slightly speeded up page loading.

**Random script suffix**
This random suffix is generated to achieve "cache busting" - any change to file content will result in different hash and hence browser is forced to load a new version of the file.
We are able to turn off this output-hashing, but then we loose benefits of mentioned cache busting.
I wasn't able to find out a way how to HTTP/2 PUSH files containing this random hash.

As I find out more, I will improve the use of HTTP/2 PUSH functionality in the YANG Catalog.

Regards,
Slavomir Mazur

**Comment by evyncke** [comment]
Hello @SlavomirMazurPantheon 

Thank you for the above, as written in another issue today, adding dynamically the Link header is a good thing.

About the random suffix, I am sure that it can be extracted in the deployment/compilation script once it is compiled. Then added in the index.html.

**Comment by evyncke** [comment]
Using per page "Link" headers is better indeed than pushing the same set of objects in all pages.

For the random suffix, it is indeed to avoid caching, but I am sure that this suffix can be recovered (at the very worse in the compile script or by scanning the resulting html file) and link can then be made

**Comment by SlavomirMazurPantheon** [comment]
Okay, I can try to come up with some workaround to get the created hash and then use it 👍
